File: occasum-core/src/main/java/pw/stamina/occasum/properties/enums/CaseInsensitiveEnumParsingService.java
package pw.stamina.occasum.properties.enums;

import pw.stamina.occasum.properties.PropertyParseException;

final class CaseInsensitiveEnumParsingService<T extends Enum<T>>
        extends AbstractEnumParsingService<T> {

    CaseInsensitiveEnumParsingService(Class<T> enumClass) {
        super(enumClass);
    }

    @Override
    public T parse(String input) throws PropertyParseException {
        if (input == null) {
            throw PropertyParseException.nullInput();
        }

        T[] constants = enumClass.getEnumConstants();

        for (T constant : constants) {
            if (constant.name().equalsIgnoreCase(input)) {
                return constant;
            }
        }

        throw new PropertyParseException("could not parse input to an enum value", input);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/enums/CaseInsensitiveEnumParsingService.kt
package pw.stamina.occasum.properties.enums

import pw.stamina.occasum.properties.PropertyParseException

internal class CaseInsensitiveEnumParsingService<T : Enum<T>>(enumClass: Class<T>)
    : AbstractEnumParsingService<T>(enumClass) {

    @Throws(PropertyParseException::class)
    override fun parse(input: String): T {
        for (constant in enumConstants) {
            if (constant.name.equals(input, ignoreCase = true)) {
                return constant
            }
        }

        throw PropertyParseException("could not parse input to an enum value", input)
    }
}